src/ExtractTileOperations.cpp

```diff
@@ -0,0 +1,322 @@
+    // ramp(x10(base), x10(stride), 3) + x6(ramp(idx, 1, 5))  y: 2, x: 3, r: 5
+    Expr first = std::move(matches[0]);
+    Expr second = std::move(matches[1]);
+    Expr adj = std::move(matches[2]);
+    const auto *r1 = first.as<Ramp>();
+    const auto *b2 = second.as<Broadcast>();
+    if (!r1 && !b2) {
+        // Try switching the order
+        r1 = second.as<Ramp>();
+        b2 = first.as<Broadcast>();
+    }
+    if (!r1 || !b2) { return {}; }
+
+    const auto *b1 = r1->base.as<Broadcast>();
+    const auto *r2 = b2->value.as<Ramp>();
+
+    if (!b1 || !r2) { return {}; }
+
+    int x_tile = r1->lanes;
+    int r_tile = r2->lanes;
+    int y_tile = b1->lanes / r_tile;
+    if (y_tile != b2->lanes / x_tile) { return {}; }
+
+    auto pattern1 = Ramp::make(Broadcast::make(wild_i32, b1->lanes), Broadcast::make(wild_i32, b1->lanes), r1->lanes);
+    if (!expr_match(pattern1, first, matches)) { return {}; }
+    Expr base = std::move(matches[0]);
+    Expr x_stride = std::move(matches[1]);
+
+    auto pattern2 = Broadcast::make(Ramp::make(wild_i32, wild_i32, r2->lanes), b2->lanes);
+    if (!expr_match(pattern2, second, matches)) { return {}; }
+    base += std::move(matches[0]);
+    Expr r_stride = std::move(matches[1]);
+
+    auto pattern3 = Broadcast::make(wild_i32, b1->lanes * r1->lanes);
+    if (!expr_match(pattern3, adj, matches)) { return {}; }
+    base -= std::move(matches[0]);
+
+    return {true, base, {x_stride, 0, r_stride}, {x_tile, y_tile, r_tile}};
+}
+
+struct NewMatmul {
+    bool result = false;
+    Stmt stmt;
+    int tile_x;
+    int tile_y;
+    int tile_r;
+};
+
+NewMatmul
+convert_to_matmul(const Store *op, const string &new_name) {
+    // m[ramp(0, 1, S)] = VectorAdd(lhs[{XYR tile}] * xX(rhs[{YR tile}])) + m[ramp(0, 1, S)]
+    const auto wild_i8x = Variable::make(Int(8, 0), "*");
+    const auto wild_u8x = Variable::make(UInt(8, 0), "*");
+    vector<Expr> matches;
+    const auto pattern1 = wild_i32x + wild_i32x;
+    if (!expr_match(pattern1, op->value, matches)) { return {}; }
+    const auto *reduce = matches[0].as<VectorReduce>();
+    const auto *load = matches[1].as<Load>();
+    if (!reduce || reduce->op != VectorReduce::Add) { return {}; }
+    if (!load || load->name != op->name || !equal(load->index, op->index)) { return {}; }
+
+    // FIXME: Add support for bf16 for LLVM 13+
+    auto pattern2 = cast(Int(32, 0), cast(Int(32, 0), wild_i8x) * wild_i32x);
+    auto pattern2_unsigned = cast(Int(32, 0), cast(Int(32, 0), wild_u8x) * wild_i32x);
+
+    bool lhs_signed = false;
+    if (expr_match(pattern2, reduce->value, matches)) {
+        lhs_signed = true;
+    } else if (expr_match(pattern2_unsigned, reduce->value, matches)) {
+        lhs_signed = false;
+    } else {
+        return {};
+    }
+
+    const auto *lhs_load = matches[0].as<Load>();
+    // FIXME: When tile_r is not 4 the broadcast is inside the index, not of the value
```

Comment:
Does this need a tracking issue?

Reply:
I don't think either of these need tracking issues, I'll remove the comment.